Source/Evolve/WarpXEvolve.cpp

```diff
@@ -217,8 +217,16 @@ WarpX::Evolve (int numsteps)
             // B : guard cells are NOT up-to-date
         }
 
-        if (cur_time + dt[0] >= stop_time - 1.e-3*dt[0] || step == numsteps_max-1) {
+        if ((cur_time + dt[0] >= stop_time - 1.e-3*dt[0] || step == numsteps_max-1) &&
+            !synchronize_velocity_for_diagnostics) {
             // At the end of last step, push p by 0.5*dt to synchronize
+            // Note that this is potentially buggy since the PushP will do a field gather
```

Comment:
Do we need a couple of input option asserts to avoid potentially buggy behavior?

Reply:
The buggy behavior regarding particle bounds is not dependent on input options, but will always happen. For the other, Python with electrostatic, this should just be fixed by removing this block of code and using the one below instead. I can do this in a separate PR because it will break a number of CI tests.